### PR TITLE
Improved filters

### PR DESCRIPTION
**Pull request by gustavlrsn**
For https://github.com/opencollective/opencollective/issues/7033

# Description
- New filtering UI and system, applied on all Dashboard pages
- `<Filterbar />` component to handle the filters and views a.k.a tabs
- `useQueryFilter`, extended this hook (and deprecated and moved an older version of this still used on some legacy admin panel pages) to handle most logic around filtering and setting values in the query etc
  - the useQueryFilter hook is configured by these parameters:
    - `schema`: a zod schema, that defines the available filters on a page and parses the values from the query
    - `toVariables`: functions to transform the filter values into API variables expected by the GraphQL query, only need to define a function here if either the key and value of the filter does not correlate to what the GraphQL query expects
    - `filters`: configurations and components for the frontend filters, which contains (or creates) the available options to the user
    - `views`: optional pre-defined tabs/views. views[0].filter will be used as the default filter unless `defaultFilterValues` is passed along. 
  
## Caveats
  - this PR essentially duplicates all dashboard pages so that there is a standalone version of each page, so that the old filters can remain in the legacy admin panel.
  - Transactions page currently lack these filters:
    - exclude gift card transactions
    - exclude pending
    - exclude children transactions
  
# Screenshots
Screenshot 2023-11-16 at 23 41 21

image
Screenshot 2023-11-16 at 23 40 31
Screenshot 2023-11-16 at 23 37 50
Screenshot 2023-11-16 at 13 58 52
New empty results screen applied across all Dashboard pages, with button to reset filter.

### CODE REVIEW

**Comment by gustavlrsn** [comment]
Current dependencies on/for this PR:
* main
  * **PR #9478** <a href="https://app.graphite.dev/github/pr/opencollective/opencollective-frontend/9478" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/opencollective/opencollective-frontend/9478?utm_source=stack-comment).

**Review by kewitz — APPROVED** [review]
Great work! I like the API for the filters component and I think it makes sense with the idea of having a consistent frontend experience of filters and tabs, while leaving the heavy work for the backend.

The only feedback I have here is that I feel like using zod in the frontend is a bit of an overkill. I really wish static type checking in the frontend combined with the backend validation GraphQL was enough.

File: package.json [hunk 1]
```diff
@@ -129,7 +134,12 @@
     "uuid": "9.0.1",
     "validator": "13.11.0",
     "webpack-manifest-plugin": "5.0.0",
-    "winston": "3.11.0"
+    "winston": "3.11.0",
+    "zod": "3.22.4"
+  },
+  "resolutions": {
+    "@radix-ui/react-dismissable-layer": "^1.0.5",
+    "@radix-ui/react-focus-scope": "^1.0.4"
```
[the remark this answers is not in the document]
Comment (gustavlrsn) [reply package.json:142]: This solves an issue of mismatched versions of these libraries for different radix ui libraries, causing focus trap issues when using for instance Dialog at the same time as a DropdownMenu.

More info here: https://github.com/radix-ui/primitives/issues/1088

**Comment by gustavlrsn** [comment]
> Great work! I like the API for the filters component and I think it makes sense with the idea of having a consistent frontend experience of filters and tabs, while leaving the heavy work for the backend.
> 
> The only feedback I have here is that I feel like using zod in the frontend is a bit of an overkill. I really wish static type checking in the frontend combined with the backend validation GraphQL was enough.

@kewitz Thank you!

For the most basic filter types zod might indeed be overkill, but I think where it shine is helping out with parsing and coercion of more complex data structures like the date filter, where the possible filter values in the frontend does not align with what GraphQL expects. 

But yeah, eventually it would be nice if the possible filter types fully aligned with what the API accepts!

**Comment by gustavlrsn** [comment]
### Merge activity

* **Nov 20, 6:40 AM**: @gustavlrsn merged this pull request with [Graphite](https://app.graphite.dev/github/pr/opencollective/opencollective-frontend/9478).

**Pull request merged**